### PR TITLE
fix: Button group widget's pop-over/drop-down enlarged along the width

### LINKED ISSUE #11564: [Bug]: Button group widget's pop-over/drop-down enlarged along the width
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When we turn the button inside a button group widget into a menu type, then the pop-over/dropdown of that menu button has an enlarged width. 

Below is the image for reference:
Screenshot 2022-03-02 at 4 43 42 PM

### Steps To Reproduce

1. Drag and drop the button group widget on to the canvas.
2. Change one of the buttons to menu type.
3. Add some menu items to it.
4. Then click on the button.
5. Observe.

### Environment

Release

### Version

Cloud

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description

- When we turn the button inside a button group widget into a menu type, then the pop-over/dropdown of that menu button has an enlarged width.
- Make the popover width to be the same as the target width

Fixes #11564 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/11564-button-group-popover-enlarged-width 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.51 **(-0.04)** | 36.99 **(-0.03)** | 35.12 **(-0.06)** | 55.81 **(-0.04)**
 :green_circle: | app/client/src/selectors/commentsSelectors.ts | 85.25 **(1.64)** | 64.71 **(2.95)** | 73.33 **(0)** | 90.59 **(2.35)**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 88.98 **(-0.78)** | 70.59 **(-1.96)** | 100 **(0)** | 92.38 **(-0.95)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :red_circle: | app/client/src/widgets/ButtonGroupWidget/component/index.tsx | 18.92 **(-6.08)** | 8.64 **(-2.56)** | 2.22 **(-1.95)** | 20.15 **(-6.59)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@jsartisan please check this code out

**Comment by jsartisan** [comment]
The solution can be improved. We need to have min-width for the dropdown

image

**Comment by jsartisan** [comment]
Just checked the issue. this issue can be solved with just CSS. We can use a combination of min-width and max-width.

**Review by dilippitchika — DISMISSED** [review]

**Comment by dilippitchika** [comment]
Screenshot 2022-03-18 at 10 30 08 AM

@wmdev0808 there is a small misalignment between the pop-over and the button, happening when I see this in the view mode. Can you please check this?

**Review by Tooluloope — CHANGES_REQUESTED** [review]

File: app/client/src/widgets/ButtonGroupWidget/component/index.tsx [hunk 1]
```diff
@@ -29,12 +29,37 @@ import {
...
+  id: string;
+  type: string;
+  label: string;
+  iconName: string;
+}
+// extract props influencing to width change
+const getButtonData = (
+  groupButtons: Record<string, GroupButtonProps>,
+): ButtonData[] => {
+  const buttonData = Object.keys(groupButtons).reduce((acc: any[], id) => {
```
Comment (Tooluloope) [inline app/client/src/widgets/ButtonGroupWidget/component/index.tsx:48]: Could we avoid the use of `any`
```
const getButtonData = (
  groupButtons: Record<string, GroupButtonProps>,
): ButtonData[] => {
  return Object.keys(groupButtons).reduce((acc: ButtonData[], id) => {
    return [
      ...acc,
      {
        id,
        type: groupButtons[id].buttonType as string,
        label: groupButtons[id].label as string,
        iconName: groupButtons[id].iconName as string,
      },
    ];
  }, []);
};
```

File: app/client/src/widgets/ButtonGroupWidget/component/index.tsx [hunk 2]
```diff
@@ -397,8 +415,142 @@ function PopoverContent(props: PopoverContentProps) {
...
+          itemWidths: Object.keys(this.props.groupButtons).reduce((acc, id) => {
+            if (this.props.groupButtons[id].buttonType === "MENU") {
+              return {
+                ...acc,
+                [id]: this.state.itemRefs[id].current?.getBoundingClientRect()
+                  .width,
+              };
+            }
+            return acc;
+          }, {}),
```
Comment (Tooluloope) [inline app/client/src/widgets/ButtonGroupWidget/component/index.tsx:461]: Could you refactor this into a function, it's similar with the function in Line 483

File: app/client/src/widgets/ButtonGroupWidget/component/index.tsx [hunk 3]
```diff
@@ -397,8 +415,142 @@ function PopoverContent(props: PopoverContentProps) {
...
+            ...this.state,
+            itemRefs: Object.keys(this.props.groupButtons).reduce((acc, id) => {
+              if (this.props.groupButtons[id].buttonType === "MENU") {
+                return {
+                  ...acc,
+                  [id]: createRef(),
+                };
+              }
+              return acc;
+            }, {}),
```
Comment (Tooluloope) [inline app/client/src/widgets/ButtonGroupWidget/component/index.tsx:540]: Also, this is similar with the reduce function in line 434

**Review by Tooluloope — CHANGES_REQUESTED** [review]
LGTM

**Review by Tooluloope — APPROVED** [review]

**Comment by shwetha-ramesh** [comment]
@wmdev0808 tested this dp. 
As per the comments the agreed design would be "Make the popover width to be the same as the target width"
But when a button group widget is dropped on a canvas the drop down is not same as the button 
image
https://loom.com/share/db4e6825306f4e369e7395b1619bd034

**Comment by dilippitchika** [comment]
@shwetha-ramesh we have added a minimum width on this so that the dropdown doesn't become too small. It follows the similar behavior as the menu button widget.

**Comment by shwetha-ramesh** [comment]
@wmdev0808 as discussed please find the below video and steps to reproduce the issue when
1. Drag and drop button group widget  on fluid width canvas > Change the orientation to VERTICAL > Check the drop down
2. Now change the canvas size to Mobile device > D&D button group widget > Check drop down > Now verify first button group widget drop down.

https://loom.com/share/3f9ef89253be41f0bb1d7e3d5720046b

**Comment by shwetha-ramesh** [comment]
@wmdev0808 re-tested this DP after fix. Working as expected.

**Review by Tooluloope — APPROVED** [review]

**Pull request merged**